subcommands/keys/rotate_root.go

```diff
@@ -1,220 +1,273 @@
-        data = json.load(f)
-    data['auth']['server'] = 'https://app.foundries.io/oauth'
-    data['auth']['client_id'] = client_id
-    data['auth']['client_secret'] = client_secret
-
-    with open(config_json, 'w') as f:
-        json.dump(data, f)
-
-with TemporaryDirectory() as tempdir:
-    os.chdir(tempdir)
-    os.mkdir('tuf')
-    creds_file = '/creds.tgz'
-    expires = (datetime.datetime.now() + datetime.timedelta(days=2*365)).strftime('%Y-%m-%dT%H:%M:%SZ')
-    cmd('tar', 'xf', creds_file, cwd='./tuf')
-    update_creds('./tuf/tufrepo/config.json', sys.argv[1], sys.argv[2])
-    cmd('garage-sign', 'root', 'pull', '--repo', './tufrepo')
-    old_keyname, old_keyid = find_current_root('./tuf/tufrepo')
-    keyname = 'offline-root-' + datetime.datetime.now().isoformat()
-    cmd('garage-sign', 'key', 'generate', '--repo', './tufrepo',
-    '--type', 'rsa', '--name', keyname)
-    cmd('garage-sign', 'root', 'key', 'add', '--repo', './tufrepo',
-    '--key-name', keyname)
-    cmd('garage-sign', 'root', 'key', 'remove', '--repo', './tufrepo',
-    '--key-id', old_keyid, '--key-name', old_keyname)
-    cmd('garage-sign', 'root', 'sign', '--repo', './tufrepo',
-    '--key-name', keyname, '--key-name', old_keyname, '--expires', expires)
-    cmd('tar', 'czf', creds_file, 'tufrepo', cwd='./tuf')
-    cmd('garage-sign', 'root', 'push', '--repo', './tufrepo')
-`
-	fmt.Println("Pulling aktualizr image...")
-	if err := pullContainer(aktualizrImageName); err != nil {
-		exitf("failed to pull image, %q, %s", aktualizrImageName, err)
+func doRotateRoot(cmd *cobra.Command, args []string) {
+	factory := viper.GetString("factory")
+	credsFile := args[0]
+	assertWritable(credsFile)
+	creds := getOfflineCreds(credsFile)
+
+	root, err := api.TufRootGet(factory)
+	subcommands.DieNotNil(err)
+
+	curid, curPk, err := findRoot(*root, creds)
+	fmt.Println("= Current root:", curid)
+	subcommands.DieNotNil(err)
+
+	// A rotation is pretty easy:
+	// 1. change the who's listed as the root key: "swapRootKey"
+	// 2. sign the new root.json with both the old and new root
+
+	newid, newPk, newCreds := swapRootKey(root, curid, creds)
+	fmt.Println("= New root:", newid)
+
+	fmt.Println("= Resigning root.json")
+	signers := []TufSigner{
+		{Id: curid, Key: curPk},
+		{Id: newid, Key: newPk},
 	}
+	subcommands.DieNotNil(signRoot(root, signers...))
+
+	bytes, err := json.MarshalIndent(root, "", "  ")
+	subcommands.DieNotNil(err)
 
-	scriptPath, err := loadScript(rotateCmd)
+	// Create a backup before we try and commit this:
+	tmpCreds := saveTempCreds(credsFile, newCreds)
+
+	fmt.Println("= Uploading rotated root")
+	body, err := api.TufRootPost(factory, bytes)
 	if err != nil {
-		exitf("failed to load script, %s", err)
-	}
-	defer os.Remove(scriptPath) // clean up
-	if err := copyFile(credentialsPath, credentialsBackupPath); err != nil {
-		exitf("failed to backup offline credentials, %s", err)
+		fmt.Println("\nERROR:", err)
+		fmt.Println(body)
+		os.Exit(1)
```

Comment:
we've made 2 not possible in this implemenation. So we really just have this odd scenario where root.json wasn't updated. My thinking was - just leave the tempfile around in case roo.json was somehow updated but we got a bad return (nginx crashed on the response). I guess we should print some guidance here: "please check root.json from server before deleting the temp file".

Reply:
https://github.com/foundriesio/fioctl/pull/81/commits/13d29fe7ef6f33f34b3e10ff38213c94eb20e459

Reply:
Ah, I see, you do things in this order:
1. save temp file
2. update root
3. rename temp file into creds file
4. update prod-root

A situation 2 is still possible if step 4 from above fails, but it doesn't apply to this logic, I see now.

Reply:
Also, as I understand we never remove old offline keys from a creds file, so this should work well.